cmd/in-place-restart-agent/main.go

```diff
@@ -0,0 +1,351 @@
+		Development: true,
+	})))
+	env := parseEnvOrDie()
+
+	// Create controller manager
+	mgr := createManagerOrDie(env)
+
+	// Setup the in-place restart agent controller
+	setupInPlaceRestartAgentOrDie(mgr, env)
+
+	// Start
+	start(mgr)
+}
+
+// createManagerOrDie creates a controller manager or exits with an error
+func createManagerOrDie(env env) ctrl.Manager {
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme: scheme,
+		// Disable metrics
+		Metrics: metricsserver.Options{
+			BindAddress: "0",
+		},
+		// Disable health probe
+		HealthProbeBindAddress: "0",
+		// Disable leader election
+		LeaderElection: false,
+		// Only watch the associated JobSet
+		// This is done to reduce the network traffic as much as possible
+		Cache: cache.Options{
+			DefaultNamespaces: map[string]cache.Config{
+				env.Namespace: {},
+			},
+			ByObject: map[client.Object]cache.ByObject{
+				&jobset.JobSet{}: {
+					Field: fields.OneTermEqualSelector("metadata.name", env.JobSetName),
+				},
+			},
+		},
+	})
+	if err != nil {
+		setupLog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	return mgr
+}
+
+// setupInPlaceRestartAgentOrDie sets up the in-place restart agent controller with the manager or exits with an error
+func setupInPlaceRestartAgentOrDie(mgr ctrl.Manager, env env) {
+	inPlaceRestartAgent := NewInPlaceRestartAgent(
+		mgr.GetClient(),
+		env.Namespace,
+		env.PodName,
+		env.WorkerCommand,
+		env.InPlaceRestartExitCode,
+	)
+	if err := inPlaceRestartAgent.SetupWithManager(mgr); err != nil {
+		setupLog.Error(err, "unable to create controller", "controller", ControllerName)
+		os.Exit(1)
+	}
+}
+
+// start starts the controller manager and waits for it to stop
+// It also handles signals to exit with an appropriate exit code like 128 + signal number instead of 0
+func start(mgr ctrl.Manager) {
+	// Create a context that is cancelled when a signal is received
+	// This is done to make the program exit with an appropriate exit code like 128 + signal number instead of 0
+	ctx, cancel := context.WithCancel(context.Background())
+	signalChannel := make(chan os.Signal, 1)
+	signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGTERM)
+	var capturedSignal os.Signal
+	go func() {
+		capturedSignal = <-signalChannel
+		setupLog.Info("received signal, shutting down", "signal", capturedSignal)
+		cancel()
+	}()
+
+	// Start manager
+	setupLog.Info("starting manager")
+	if err := mgr.Start(ctx); err != nil {
+		setupLog.Error(err, "problem running manager")
+		os.Exit(1)
+	}
+
+	// Exit with 128 + signal number if the manager stopped due to a signal
+	if capturedSignal != nil {
+		// If received a signal, exit with 128 + signal number
+		if signal, ok := capturedSignal.(syscall.Signal); ok {
+			os.Exit(128 + int(signal))
+		}
+		// Fallback if casting fails (unlikely for SIGINT/SIGTERM)
+		os.Exit(1)
+	}
+
+	// If the manager stopped without an error and without a signal, exit with 0
+	os.Exit(0)
```

Comment:
Maybe we should add some log statements before existing the manager with various exist codes?